src/cmdline_manager.ts

```diff
@@ -1,128 +1,208 @@
+            commands.registerCommand("vscode-neovim.test-cmdline", this.testCmdline),
+        );
     }
 
     public dispose() {
-        this.commandLine?.dispose();
         disposeAll(this.disposables);
     }
 
+    private reset() {
+        this.lastTypedText = "";
+        this.ignoreAcceptEvent = false;
+        this.ignoreHideEvent = false;
+        this.redrawExpected = true;
+        this.pendingNvimUpdates = 0;
+        this.input.value = "";
+        this.input.title = "";
+        this.input.items = [];
+        this.input.activeItems = [];
+    }
+
     private handleRedraw({ name, args }: EventBusData<"redraw">) {
         switch (name) {
             case "cmdline_show": {
-                const [content, _pos, firstc, prompt, _indent, _level] = args[0];
-                const allContent = content.map(([, str]) => str).join("");
-                // !note: neovim can send cmdline_hide followed by cmdline_show events
-                // !since quickpick can be destroyed slightly at later time after handling cmdline_hide we want to create new command line
-                // !controller and input for every visible cmdline_show event
-                // !otherwise we may hit cmdline_show when it's being hidden
-                // as alternative, it's possible to process batch and determine if we need show/hide or just redraw the command_line
-                // but this won't handle the case when cmdline_show comes in next flush batch (is it possible?)
-                // btw, easier to just recreate whole command line (and quickpick inside)
-                if (this.cmdlineTimer) {
-                    clearTimeout(this.cmdlineTimer);
-                    this.cmdlineTimer = undefined;
-                    this.showCmd(allContent, firstc, prompt);
+                const [icontent, _pos, firstc, prompt, _indent, _level] = args[0];
+                const content = icontent.map(([, str]) => str).join("");
+                logger.debug(`cmdline_show: "${content}"`);
+                this.lastTypedText = content;
+                this.input.title = prompt || this.getTitle(firstc);
+                // only redraw if triggered from a known keybinding. Otherwise, delayed nvim cmdline_show could replace fast typing.
+                if (!this.redrawExpected) {
+                    logger.debug(`cmdline_show: ignoring cmdline_show because no redraw expected: "${content}"`);
+                    break;
+                }
+                this.redrawExpected = false;
+                this.input.show();
+                if (this.input.value !== content) {
+                    this.pendingNvimUpdates++;
+                    const activeItems = this.input.activeItems; // backup selections
+                    this.input.value = content; // update content
+                    this.input.activeItems = activeItems; // restore selections
+                }
+                break;
```

Comment:
Nitpick: I might move all of this into its own method, just to keep this switch statement slim